### DISCUSSION TITLE
Error: "Unsupported board: sw0 devicetree alias is not defined" in Button Sample

### DISCUSSION THREAD

**Discussion by jayesh0711** [zephyrproject-rtos/zephyr#81266] [General] [answered]
Message:

Hi Zephyr Community,

I'm building the button sample for the mec15xxevb_assy6853 board, but I encountered the following error:

/home/jayesh/zephyrproject/zephyr/samples/basic/button/src/main.c:117:2: error: #error "Unsupported board: sw0 devicetree alias is not defined"

It seems like the sw0 devicetree alias is not defined in the configuration. How can I resolve this issue or define the alias for the button (sw0) in the devicetree?

Any guidance would be greatly appreciated!

**Comment by iamadahn** [comment]
Hello, sw0 is not defined in devicetree for this board, so you have to create an overlay and define a button and it's alias yourself.
You can learn how to do that from this guide - https://docs.zephyrproject.org/latest/build/dts/howtos.html
The section you must be looking for is "Set devicetree overlays".

**Comment by leenaarora18** [comment]
Hi,
I am working on the MEC15XXEVB_ASSY6853 board and have defined LEDs in my devicetree as follows:
leds {
    compatible = "gpio-leds";
    led2: led_0 {
        gpios = < &gpio_140_176 0xe 0x1 >;
        label = "LED 2";
    };
    led3: led_1 {
        gpios = < &gpio_140_176 0xf 0x1 >;
        label = "LED 3";
    };
    led4: led_2 {
        gpios = < &gpio_140_176 0xb 0x1 >;
        label = "LED 4";
    };
};
I understand that gpio_140_176 represents a GPIO controller handling a range of GPIO pins, but I am unclear how the pin numbers like 0xe, 0xf, and 0xb are determined. Are these offsets within the GPIO controller range, or do they represent specific mappings from the board's hardware schematic?

I have already cross-checked with the MEC1501 datasheet but couldn’t find clear information about these values or how they relate to the board's GPIO configuration.

**Reply by iamadahn** [reply]
Have you tried looking into official Zephyr dts for this board?
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/microchip/mec15xxevb_assy6853/mec15xxevb_assy6853.dts
As I understand there are macros for decoding pins, for example "MCHP_GPIO_DECODE_156". 
To initialize a led they provide only this macro and gpio active level.

**Reply by leenaarora18** [reply]
but i am asking from where you are seeing this 0xe, 0xf, and 0xb  basically this are the pin mapping to respective led but from where you are seeing this. i already check in code there is no define for this and also in datasheet and schematic i am not able to see this please help .

**Reply by iamadahn** [reply, marked as answer]
According to zephyr/include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h file, for example `MCHP_GPIO_DECODE_156 `goes to `XEC_GPIO_HELPER(&gpio_140_176, 14)` which goes to `&gpio_140_176, 14` which can be rewritten as  `&gpio_140_176, 0xE`. So `MCHP_GPIO_DECODE_156` macro means `&gpio_140_176, 14`.

**Comment by leenaarora18** [comment]
Hi Dude I am working on the MEC1521 ASSY 6883 development board and have implemented a simple application to toggle an LED when a button is pressed. I am following the zephyr/include/zephyr/dt-bindings/gpio/microchip-xec-gpio.h file for GPIO macros, and here is an outline of my setup:

**1.**Main Code (main.c)****
#include <zephyr/kernel.h>
#include <zephyr/device.h>
#include <zephyr/drivers/gpio.h>
#include <zephyr/sys/printk.h>
#include <inttypes.h>

#define SLEEP_TIME_MS 1

#define SW0_NODE DT_ALIAS(sw0)
#define LED0_NODE DT_ALIAS(led1)

#if !DT_NODE_HAS_STATUS(SW0_NODE, okay)
#error "Unsupported board: button alias 'sw0' not defined in the device tree"
#endif

#if !DT_NODE_HAS_STATUS(LED0_NODE, okay)
#error "Unsupported board: LED alias 'led0' not defined in the device tree"
#endif

static const struct gpio_dt_spec button = GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios, {0});
static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET_OR(LED0_NODE, gpios, {0});

static struct gpio_callback button_cb_data;

void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
{
    printk("Button pressed on %s at %" PRIu32 "\n", dev->name, k_cycle_get_32());
    gpio_pin_toggle_dt(&led);
}

int main(void)
{
    int ret;
    printk("Hello Surjit\n");

    if (!device_is_ready(led.port)) {
        printk("Error: LED device %s is not ready\n", led.port->name);
        return -1;
    }
    ret = gpio_pin_configure_dt(&led, GPIO_OUTPUT_ACTIVE);
    if (ret != 0) {
        printk("Error %d: failed to configure LED %s pin %d\n", ret, led.port->name, led.pin);
        return ret;
    }

    if (!device_is_ready(button.port)) {
        printk("Error: Button device %s is not ready\n", button.port->name);
        return -1;
    }
    ret = gpio_pin_configure_dt(&button, GPIO_INPUT);
    if (ret != 0) {
        printk("Error %d: failed to configure button %s pin %d\n", ret, button.port->name, button.pin);
        return ret;
    }
    ret = gpio_pin_interrupt_configure_dt(&button, GPIO_INT_EDGE_TO_ACTIVE);
    if (ret != 0) {
        printk("Error %d: failed to configure interrupt for button %s pin %d\n", ret, button.port->name, button.pin);
        return ret;
    }

    gpio_init_callback(&button_cb_data, button_pressed, BIT(button.pin));
    gpio_add_callback(button.port, &button_cb_data);
    printk("Set up button on %s pin %d\n", button.port->name, button.pin);
    printk("Set up LED on %s pin %d\n", led.port->name, led.pin);

    while (1) {
        k_msleep(SLEEP_TIME_MS);
    }

    return 0;
}
**2.Device Tree Overlay (board.overlay):**
/ {
    aliases {
        sw0 = &button0;
    };
    buttons {
        compatible = "gpio-keys";
        button0: button_0 {
            gpios = <&gpio_140_176 0x12 GPIO_ACTIVE_HIGH>;
            label = "User Button";
        };
    };
};
**3.Problem:**

When I run the code, I can see "Button pressed" messages in the logs, but the interrupt fires continuously even when the button is pressed and released once.
Logs: 
Hello Surjit
Set up button on gpio@40081180 pin 18
SeButton pressed on gpio@40081180 at 317
t up LED on gpio@40081180 pin 15
Button pressed on gpio@40081180 at 980
Button pressed on gpio@40081180 at 1638
Button pressed on gpio@40081180 at 2298
Button pressed on gpio@40081180 at 2956
Button pressed on gpio@40081180 at 3614
Button pressed on gpio@40081180 at 4275
Button pressed on gpio@40081180 at 4933
Button pressed on gpio@40081180 at 5592
Button pressed on gpio@40081180 at 6249

**Questions:**

Why is the interrupt firing continuously instead of only once per button press?
Is my device tree overlay correct for the switch-S4
![Screenshot 2024-11-19 192537](https://github.com/user-attachments/assets/5241a7d7-0477-4280-8c17-dcece3c106f3)
![Screenshot 2024-11-19 192537](https://github.com/user-attachments/assets/59c943fa-6cfa-4201-aa14-f2070bb12e79)
![Screenshot 2024-11-19 192537](https://github.com/user-attachments/assets/9928cd06-4c02-4cdf-ad6e-c5f529d964a2)
 configuration using gpio_140_176?
Are there any additional debounce or interrupt settings required for the MEC1521 GPIOs?
Any guidance would be greatly appreciated. Thank you!

**Reply by iamadahn** [reply]
I think that there might be problem in your connection, according to the scheme you provided, only 1 set of jumpers must be set at one time, either JP17 or JP23, so check it out.

**Reply by leenaarora18** [reply]
Thanks Its working now.

**Reply by iamadahn** [reply]
You are welcome.

**Comment by jayesh0711** [comment]
Hi adahn

I'm encountering errors when trying to build my Zephyr application for controlling an LED using PWM. and i am using the mec1521 assy 6883 board only .

Below are the details of my configuration:
Code Overview:
In the main.c file, I’m trying to control an LED using PWM, and here is the relevant code:
#include <zephyr/kernel.h>
#include <zephyr/sys/printk.h>
#include <zephyr/device.h>
#include <zephyr/drivers/pwm.h>
/* Define PWM specifications for the LED */
static const struct pwm_dt_spec pwm_led0 = PWM_DT_SPEC_GET(DT_ALIAS(pwm_led0));
#define MIN_PERIOD PWM_MSEC(1)
#define MAX_PERIOD PWM_SEC(1U)
void main(void) {
printk("PWM LED Control Example\n");
if (!pwm_is_ready_dt(&pwm_led0)) {
printk("Error: PWM device is not ready\n");
return;
}
uint32_t period = MAX_PERIOD;
uint8_t dir = 0; /* Direction of change: 0 = decrease, 1 = increase */
while (1) {
/* Set PWM for the LED */
pwm_set_dt(&pwm_led0, period, period / 2);
printk("PWM Period: %u\n", period);
/* Adjust period */
if (dir) {
period *= 2;
if (period > MAX_PERIOD) {
period = MAX_PERIOD / 2;
dir = 0;
}
} else {
period /= 2;
if (period < MIN_PERIOD) {
period = MIN_PERIOD * 2;
dir = 1;
}
}
k_sleep(K_MSEC(500)); /* Adjust delay as needed */
}
}

Device Tree Configuration:
Here is the relevant section from my .dts overlay:
& pwm-0 {
status = "okay";
channel-gpios = <&gpio_040_076 0xb GPIO_ACTIVE_HIGH>;
};
pwmleds {
compatible = "pwm-leds";
pwm_led0: pwm_led_0 {
pwms = <&pwm-0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>; /* Channel 0, Period 20ms */
label = "PWM_LED0";
};
};

/* Define PWM channel for LED2 */
& pwm_led0 {
pwms = <&pwm-0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>; /* Channel 0 */
label = "PWM_LED2";
};

Errors Encountered:
When trying to build the project, I am receiving the following errors:
1. error: 'DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_VAL_period' undeclared
here (not in a function)
2. error: 'DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_VAL_channel' undeclared
here (not in a function)
3. error: '__device_dts_ord_DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_PH_ORD'
undeclared here (not in a function)
The error seems to be related to how the PWM alias and associated properties are being handled.
What I’ve Tried:
 Verified the device tree overlay configuration.
 Ensured that the aliases are correctly defined.
 Checked if the PWM driver is correctly registered.
Question:
Could anyone help me understand why these undeclared errors are happening? Is there an issue with how
I have defined the alias for the PWM LED, or am I missing something in the device tree setup? Any
guidance on fixing these errors would be appreciated.
Thanks in advance!

**Reply by iamadahn** [reply]
Hello, could you provide a link for repository you are working with so I can try to figure out the problem?

**Reply by jayesh0711** [reply]
> Hello, could you provide a link for repository you are working with so I can try to figure out the problem?

https://github.com/leenaarora18/Blinky_pwm-mec1521-6883

**Reply by iamadahn** [reply]
For some reason you have a lot of .overlay files in `boards` folder, but there is no needed .overlay file for your specific board, so west is trying to build the project with default .dts file from zephyr sources and there is no `pwm_led0` provided.
I suggest you to remove all unneeded .overlay files from `boards` folder and instead provide app.overlay file with needed pwm node and alias for it.

**Reply by leenaarora18** [reply]
Hii, i have done the necessary changes as you suggested. Can you please look in to this and suggest a solution for the same. link is share below:
https://github.com/leenaarora18/Blinky_pwm-mec1521-6883.git

**Comment by jayesh0711** [comment]
I recently updated the repository with the suggestions provided, and now when I try to build the application for the mec15xxevb_assy6853 board, I'm encountering the following error:


CMake Error at /home/jayesh/zephyrproject/zephyr/cmake/modules/dts.cmake:311 (message):
  gen_defines.py failed with result code: 1 - stderr contents:

  devicetree error:
  /home/jayesh/zephyrproject/zephyr/samples/basic/blinky_pwm/boards/mec15xxevb_assy6853_mec1501_hsz.overlay:46
  (column 1): parse error: expected '/' or label reference (&foo)

Call Stack (most recent call first):
  /home/jayesh/zephyrproject/zephyr/cmake/modules/zephyr_default.cmake:132 (include)
  /home/jayesh/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/jayesh/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:4 (find_package)





The error points to a parse issue in the mec15xxevb_assy6853_mec1501_hsz.overlay file at line 46. The expected format seems to be a / or label reference, but I'm not sure what went wrong in the overlay configuration.

Has anyone faced a similar issue? Could you suggest possible solutions or guide me in troubleshooting this parse error?

Here is the updated repository link: https://github.com/leenaarora18/pwm_updated_version_1.git

Thank you for your help!

**Reply by iamadahn** [reply]
To get rid of this error you should add `&` symbol before `pwmleds`, so it will be `&pwmleds`.

**Reply by jayesh0711** [reply]
> To get rid of this error you should add `&` symbol before `pwmleds`, so it will be `&pwmleds`.

When using the &pwmleds as you suggested   configuration in the devicetree overlay:
&pwm0 {
    status = "okay";
};

&pwmleds {
    compatible = "pwm-leds";

    pwm_led0: pwm_led_0 {
        pwms = <&pwm0 0 20000000 0>; /* Channel 0, Period 20ms, Normal Polarity */
        label = "PWM_LED0";
    };
};

I'm encountering this error during the build process:
devicetree error:
.../mec15xxevb_assy6853_mec1501_hsz.overlay:62 (column 1): parse error: undefined node label 'pwmleds'

The full build log shows:

CMake fails during gen_defines.py execution.
It reports the node label pwmleds as undefined.

**Comment by iamadahn** [comment]
It's very hard to follow your code, you have 2 files with the same name - `mec15xxevb_assy6853_mec1501_hsz.overlay` in the root folder and in the boards folder and neither of them have 62 to lines, but you are saying that you have an error on line 62:

> I'm encountering this error during the build process:
devicetree error:
.../mec15xxevb_assy6853_mec1501_hsz.overlay:62 (column 1): parse error: undefined node label 'pwmleds'

**Reply by jayesh0711** [reply]
> It's very hard to follow your code, you have 2 files with the same name - `mec15xxevb_assy6853_mec1501_hsz.overlay` in the root folder and in the boards folder and neither of them have 62 to lines, but you are saying that you have an error on line 62:
> 
> > I'm encountering this error during the build process:
> > devicetree error:
> > .../mec15xxevb_assy6853_mec1501_hsz.overlay:62 (column 1): parse error: undefined node label 'pwmleds'

CMake Error at /home/jayesh/zephyrproject/zephyr/cmake/modules/dts.cmake:311 (message):

  gen_defines.py failed with result code: 1 - stderr contents:

  devicetree error:

  /home/jayesh/zephyrproject/zephyr/samples/basic/blinky_pwm/boards/mec15xxevb_assy6853.overlay:19

  (column 1): parse error: undefined node label 'pwmleds'


For reference, here is the updated link to the repository:https://github.com/leenaarora18/plinky_pwm_version_1.git

**Reply by jayesh0711** [reply]
> It's very hard to follow your code, you have 2 files with the same name - `mec15xxevb_assy6853_mec1501_hsz.overlay` in the root folder and in the boards folder and neither of them have 62 to lines, but you are saying that you have an error on line 62:
> 
> > I'm encountering this error during the build process:
> > devicetree error:
> > .../mec15xxevb_assy6853_mec1501_hsz.overlay:62 (column 1): parse error: undefined node label 'pwmleds'

Hello dahn

**Reply by iamadahn** [reply]
> /home/jayesh/zephyrproject/zephyr/samples/basic/blinky_pwm/boards/mec15xxevb_assy6853.overlay:19

According to that line, you are trying to build the blinky sample, not your project, so the compiler has no idea about .overlay file that you provide in your project

**Comment by jayesh0711** [comment]
I am facing the following error while building the Zephyr blinky_pwm sample application:

from /home/jayesh/zephyrproject/zephyr/samples/basic/blinky_pwm/src/main.c:14:
/home/jayesh/zephyrproject/zephyr/include/zephyr/device.h:92:41: error: '__device_dts_ord_DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_PH_ORD' undeclared here (not in a function)
   92 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
/home/jayesh/zephyrproject/zephyr/include/zephyr/devicetree.h:240:32: error: 'DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_VAL_channel' undeclared here (not in a function); did you mean 'DT_N_S_pwmleds_S_pwm_led_0_P_pwms_IDX_0_VAL_channel'?
  240 | #define DT_ALIAS(alias) DT_CAT(DT_N_ALIAS_, alias)
/home/jayesh/zephyrproject/zephyr/include/zephyr/devicetree.h:240:32: error: 'DT_N_ALIAS_pwm_led0_P_pwms_IDX_0_VAL_period' undeclared here (not in a function); did you mean 'DT_N_S_pwmleds_S_pwm_led_0_P_pwms_IDX_0_VAL_period'?
  240 | #define DT_ALIAS(alias) DT_CAT(DT_N_ALIAS_, alias)

I have ensured that the project configuration includes CONFIG_PWM=y, but the errors persist. How can I resolve this issue? Is there anything specific that needs to be added to the device tree or overlay file to define the pwm_led0 alias correctly?

For reference, here is the updated link to the repository: https://github.com/leenaarora18/plinky_pwm_version_2.git

**Reply by iamadahn** [reply]
You need no create an alias for your pwmled, something like that:

`
aliases {
		pwm-led0 = &pwm_led0
	};
`

And then change the code in your source file from `static const struct pwm_dt_spec pwm_led0 = PWM_DT_SPEC_GET(DT_ALIAS(pwm_led0));` to `static const struct pwm_dt_spec pwm_led0 = PWM_DT_SPEC_GET(DT_ALIAS(pwm-led0));` . This change is required because as I remember you cant use `_` symbol in alias naming.

**Comment by jayesh0711** [comment]
Can you suggest me please ?

**Reply by iamadahn** [reply]
Done.

**Reply by jayesh0711** [reply]
> Done.

Thanks its working now

**Reply by iamadahn** [reply]
You are welcome.

**Comment by leenaarora18** [comment]
Hi adahn ,

Issue with Intel Raptor Lake and MEC1521 SPI Image Flashing:

I'm working with Intel Raptor Lake and using the open-source ECFW Zephyr (https://github.com/intel/ecfw-zephyr) to generate the ksc.bin file. Here’s what I did:

Cloned the CPGZephyrDocs repository and added the necessary SPI image generation files.

Set the environment variable for everglades_spi_gen_RomE.

Successfully built the project using west build -c -p auto -b mec1501_mtl_p, which generated the ksc.bin file.

Stitched the ksc.bin with Intel BIOS and flashed it using Dediprogrammer onto the Intel Raptor Lake S board.

After flashing, both the CPU fan and ATX fan started working, and the power sequence voltages were active, but the system did not boot. However, these signals (CPU fan, ATX fan, and power sequence) only activate when I manually touch the RSMRST signal using a multimeter.

Any insights on why the system isn't booting after flashing?

**Reply by iamadahn** [reply]
Hello,

I'm afraid that I can't help with these because I don't have an appropriate experience unfortunately.
